### ISSUE TITLE
MLCS2k2

### ISSUE THREAD

**Issue by kbarbary** [sncosmo/sncosmo#81]
Implementing the [MLCS2k2 model](http://www.physics.rutgers.edu/~saurabh/mlcs2k2/) keeps coming up, so I'm opening an issue for it.

To really take advantage of all the sncosmo fitting and plotting, we'll have to somehow make a parameterized spectral template out the the model. A less ambitious goal would be just to provide I/O functions or perhaps a "model" separate from the typical sncosmo Model classes. I'm not so sure about doing a separate "model" though - seems like this might cause more confusion than its worth.
Personally I'm not that familiar with the model, but if someone is keen to do this, speak up here!

To get started, here is a short script that @drhay53 provided to parse the template files:

```python
from os.path import join
import numpy as np


def parse_mlcs_files(dirname):
    """Parse MLCS files in directory, returning a dictionary of dictionaries.

    Examples
    --------
    >>> d = parse_mlcs_files("./lcgen")
    """

    # assume specific file names
    alphabeta = join(dirname, 'alphabeta.dat')
    Uvec = join(dirname, 'vectors_U.dat')
    Bvec = join(dirname, 'vectors_B.dat')
    Vvec = join(dirname, 'vectors_V.dat')
    Rvec = join(dirname, 'vectors_R.dat')
    Ivec = join(dirname, 'vectors_I.dat')
    zeta = join(dirname, 'zeta.dat')

    # setup nested dict for each filter
    filters = 'UBVRI'
    d = {}
    for f in filters:
        d[f] = {}

    # alphabeta
    fope = open(alphabeta, 'r')
    for line in fope:
        if line[0] != '#':
            ln_spl = line.split()
            d[ln_spl[0]]['alpha'] = float(ln_spl[1])
            d[ln_spl[0]]['beta'] = float(ln_spl[2])
    fope.close()
    
    # vectors
    U = np.genfromtxt(Uvec, dtype=None, names=True)
    d['phase_vectors'] = U['phase']
    d['U']['M0'] = U['M0_U']
    d['U']['P'] = U[ 'P_U']
    d['U']['Q'] = U[ 'Q_U']

    B = np.genfromtxt(Bvec, dtype=None, names=True)        
    d['B']['M0'] = B['M0_B']
    d['B']['P'] = B[ 'P_B']
    d['B']['Q'] = B[ 'Q_B']

    V = np.genfromtxt(Vvec, dtype=None, names=True)        
    d['V']['M0'] = V['M0_V']
    d['V']['P'] = V['P_V']
    d['V']['Q'] = V['Q_V']

    R = np.genfromtxt(Rvec, dtype=None, names=True)        
    d['R']['M0'] = R['M0_R']
    d['R']['P'] = R['P_R']
    d['R']['Q'] = R['Q_R']

    I = np.genfromtxt(Ivec, dtype=None, names=True)        
    d['I']['M0'] = I['M0_I']
    d['I']['P'] = I[ 'P_I']
    d['I']['Q'] = I[ 'Q_I']
    
    z = np.genfromtxt(zeta, dtype=None, names=True)
    d['phase_zeta'] = z['phase']
    d['U']['zeta'] = z['zeta_U']
    d['V']['zeta'] = z['zeta_B']
    d['B']['zeta'] = z['zeta_V']
    d['R']['zeta'] = z['zeta_R']
    d['I']['zeta'] = z['zeta_I']
    
    return d


def lc(dirname, mu, delta, Av, Rv):
    """Return a dictionary containing 6 arrays: 'phase', 'U', 'B', 'V',
    'R', 'I'. These are the phases and the lightcurve in magnitudes in five
    filters."""

    mlcs = parse_mlcs_files(dirname)
    filters = 'UBVRI'
    lcd = {}
    lcd['phase'] = mlcs['phase_vectors']
    
    # zeta runs from a different range, cut it off
    zeta_mask = (mlcs['phase_zeta'] >= -10.) & (mlcs['phase_zeta'] <= 90.)

    for f in filters:
        lcd[f] = (mlcs[f]['M0'] + mu + mlcs[f]['zeta'][zeta_mask] *
                  (mlcs[f]['alpha'] + mlcs[f]['beta'] / Rv) * Av +
                  mlcs[f]['P']*delta + mlcs[f]['Q']*delta**2)

    return lcd
    
if __name__ == '__main__':
    import matplotlib.pyplot as plt

    lightcurve = lc("./lcgen", 40, 0, 0, 2.23)
    plt.plot(lightcurve['phase'], lightcurve['U'], color='cyan')
    plt.plot(lightcurve['phase'], lightcurve['B'], color='blue')
    plt.plot(lightcurve['phase'], lightcurve['V'], color='purple')
    plt.plot(lightcurve['phase'], lightcurve['R'], color='orange')
    plt.plot(lightcurve['phase'], lightcurve['I'], color='red')
    plt.gca().invert_yaxis()
    plt.show()
```

**Comment by drhay53** [comment]
The SNANA $SNDATA_ROOT directory has different 'versions' of MLCS, which basically translates to different sets of training vectors. One thing I noticed is that the SNANA vector files generally don't have headers, and the 'zeta' vector sometimes does start at -20 instead of -10. 

Also, on Saurabh's website, there is an 'observer-frame fitter' (http://www.physics.rutgers.edu/~saurabh/mlcs2k2/). I'm not sure exactly how it works still, but it must be assuming some spectral model. 

We should probably contact Saurabh relatively soon and open a discussion about how we might approach this in an sncosmo-like way.

**Comment by kbarbary** [comment]
@drhay53 Yeah I don't exactly understand the zeta thing, but I suspect it would be better to read or infer the range from the file rather than hard-coding it, if at all possible.

I noticed the "oberver-frame fitter" but haven't looked at it. Should be pretty easy to read the source for that and figure out what it's doing - it's IDL.

I wanted to `@`-mention Saurabh here, too bad he's not on github! :)

**Comment by rbiswas4** [comment]
I was wondering if there has been more thought along these lines. 

One of the things I am looking at with N. Karpenka (I don't know her github user name ... will add an @ later when I find out) is regarding how to incorporate her light-curve based models of core-collapse SN into SNCosmo. 

This has the ability to return band fluxes (in physical units like ergs/cm^2/s/Ang) as a function of time in given filters. This is a bit different from our concept of a model and a source (aside from the fact that it is not a Time-Series of spectra).


Here is my current very preliminary plan (comments, insights and pointers to possible problems and already done work very welcome :)   :

I am thinking of refactoring this into a Model-like class (say ModeLC) with an amplitude and redshift as parameters. In a simulation, the amplitude will describe observer frame amplitudes of band fluxes, just like 'x0' in the SALT model describes spectra on the the surface of the earth, and will involve 
cosmology, and a rest frame distribution of amplitudes'

I am planning the skipping of the step with a source here, mostly because part of her work has been to define better methods of interpolation, and I would like to see those advantages compared to a simple interpolation scheme.

**Comment by kbarbary** [comment]
Offline discussion with Saurabh resulted in the conclusion that he could possibly provide a MLCS2k2-like *spectral* template that would be similar to the result of the combination of MLCS2k2 light curve templates + k-corrections. No word on that yet, but I've been meaning to get back to him about it.

If I understand correctly, these new templates are defined for some given set of bandpasses. Which bandpasses are they? How do you plan to translate the LC in the template bandpasses to other bandpasses (or the same bandpasses at different redshifts)?

**Comment by nkarp** [comment]
There are 2 main differences between the "original" templates and my. 1st is in the way I plan to create photometry for the epochs with no spectra at a desired redshift and passband. 2nd is that I do not mix spectra from the different objects into one template. 
Answers for your questions:
"If I understand correctly, these new templates are defined for some given set of bandpasses." No, one can do it for any filter for which we have a spectra. Templates are still a set of spectra.
"How do you plan to translate the LC in the template bandpasses to other bandpasses (or the same bandpasses at different redshifts)?" 
Same as before, one takes a mangled spectra and redshift it to a desired z. But for the epochs with no spectra I will not interpolate between spectra to get a desired flux, but fit a function to flux points we have from spectral integration and get all necessary fluxes from that fit. 
I hope it is more clear now. If not, I am happy to have another attempt to explain. :)

**Comment by kbarbary** [comment]
Ah, so the difference is all in how one interpolates: It sounds like you are integrating, then interpolating, rather than interpolating then integrating? Is there a quick explanation why this is preferable for these templates? Noisy spectra perhaps?

Anyway, @rbiswas4 is right that one cannot use the existing `Model` class for this sort of light curve calcuation; it has a fundamentally different approach to integration.

**Comment by nkarp** [comment]
You got the idea right. 
I prefer to do it that way since I do not want to mix spectra from different objects within a type. Thus, if your interpolate spectra which are 10 or more days apart, or have a noisy spectra, one can get "unreasonable" fluxes.

**Comment by kbarbary** [comment]
I see. I understand not wanting to mix spectra from different objects.

Whether it is better to integrate then interpolate must depend on the interpolation method being used. For example, for simple linear interpolation, interpolation-then-integration should be exactly equivalent to integration-then-interpolation. However, I don't have a good intuition for other interpolation schemes.

**Comment by nkarp** [comment]
"for simple linear interpolation, interpolation-then-integration should be exactly equivalent to integration-then-interpolation" - correct. However, I am not planing to do linear interpolation between an integrated fluxes. I will interpolate with a function which has a "light curve" shape.

**Comment by saurabhwjha** [comment]
Hi all. I haven't been following this thread until now, and am not 100% sure if this is on the track these comments have taken, but it is based on my understanding from a few months ago.

I've created an SED based on the MLCS2k2 model (the so-called 'tweaked' model, which includes a handmade extension from -10 to -20 days). The "template" SED is just the Hsiao model, which is "mangled" to match the MLCS2k2 model UBVRI photometry. 

The results are on Dropbox at https://www.dropbox.com/sh/5l0n8aer3mafqj0/AAClAvNiUYqqqiVUc5eioow0a?dl=0
There are IDL save files (sorry) and text files (sorry), but see below.

The main result is a three-dimensional SED array of fluxes (erg/cm2/s/A) over phase (days), Delta, and wavelength (Angstroms). These are densely sampled enough such that simple interpolation for any phase (every day from -20 to +85) and Delta (in steps of 0.025 from -0.5 to 1.8, which goes beyond the range where the model is useful) to give an SED for any phase and Delta. Like Hsiao, the SED goes from 1000-25000 A, but this is well beyond where the model is defined (UBVRI only, so from ~3200-9000 A or so).

The SED is in the supernova rest-frame, and does not include any extinction. So to get the full MLCS2k2 model, we will need to apply an extinction law to the SEDs. The simplest (which is what MLCS2k2 does) is just to apply a CCM (really Fitzpatrick+Massa) law with two parameters A_V and R_V. I think Kyle said sncosmo can do this itself.

From that it should be straightforward to create MLCS2k2 model magnitudes for any t0,Delta,A_V,R_V,mu at any redshift in any observer-frame filters with some synthetic photometry routines (included in sncosmo?).

Here's what the IDL modelflux save file contains:
DELTA           DOUBLE    = Array[93]
FILTERS         STRING    = Array[5]
GOODLAMBDA      DOUBLE    = Array[2]
LAMBDA          DOUBLE    = Array[2401]
MODELFLUX       DOUBLE    = Array[93, 106, 2401]
MODELMAGS       DOUBLE    = Array[5, 93, 106]
NDELTA          LONG      =           93
NFILTERS        LONG      =            5
NLAMBDA         LONG      =         2401
NPHASE          LONG      =          106
PHASE           DOUBLE    = Array[106]

I was able to read this easily into python with:
from scipy.io import readsav
data = readsav("mlcs2k2.modelflux.sav")

I think this should be useful. Let me know where we go from here.

**Comment by kbarbary** [comment]
This sounds great. I think the next step will be to create a class. We'll chat this week! (We're both at Astro Hack Week this week).

sncosmo already has an implementation of the Fitzpatrick (1999) extinction law. We don't have Fitzpatrick and Massa (2007) but we can add it if we need to.

**Comment by saurabhwjha** [comment]
Okay, thanks to considerable help from @kbarbary, my AstroHackWeek project has made good progress and my fork of the sncosmo repository has an MLCS2k2 source you can use: https://github.com/saurabhwjha/sncosmo

Kyle and I made a demo of this; check out the IPython notebook at https://github.com/saurabhwjha/sncosmo-mlcs2k2-demo

I think it is now in a state for Kyle to ingest and merge into the main branch. 
The mlcs2k2.modelflux.fits file is now also hosted at 
http://www.physics.rutgers.edu/~saurabh/xfer/mlcs2k2.modelflux.fits

**Comment by kbarbary** [comment]
To keep others on the same page, I'm planning to give saurabh a PR to his fork with the machinery for auto-downloading and loading the model! Hopefully early next week. Then we'll merge it into sncosmo master.

**Comment by kbarbary** [comment]
Closed by #113
